sphericart/include/cuda_cache.hpp

```diff
@@ -0,0 +1,432 @@
+#include <cxxabi.h>
+
+#include "dynamic_cuda.hpp"
+
+// Helper function to demangle the type name if necessary
+std::string demangleTypeName(const std::string& name) {
+#if defined(__GNUC__) || defined(__clang__)
+    int status = 0;
+    std::unique_ptr<char, void (*)(void*)> demangled_name(
+        abi::__cxa_demangle(name.c_str(), 0, 0, &status), std::free
+    );
+    return (status == 0) ? demangled_name.get() : name;
+#else
+    throw std::runtime_error("demangling not supported using this toolchain.");
+#endif
+}
+
+// Base case: No template arguments, return function name without any type information
+std::string getKernelName(const std::string& fn_name) { return fn_name; }
+
+// Function to get type name of a single type
+template <typename T> std::string typeName() { return demangleTypeName(typeid(T).name()); }
+
+// Variadic template function to build type list
+template <typename T, typename... Ts> void buildTemplateTypes(std::string& base) {
+    base += typeName<T>(); // Add the first type
+    // If there are more types, add a comma and recursively call for the remaining types
+    if constexpr (sizeof...(Ts) > 0) {
+        base += ", ";
+        buildTemplateTypes<Ts...>(base); // Recursively call for the rest of the types
+    }
+}
+
+// Helper function to start building the types
+template <typename T, typename... Ts> std::string buildTemplateTypes() {
+    std::string result;
+    buildTemplateTypes<T, Ts...>(result); // Use recursive variadic template
+    return result;
+}
+
+/*
+Function to get the kernel name with the list of templated types if any:
+*/
+template <typename T, typename... Ts> std::string getKernelName(const std::string& fn_name) {
+    std::string type_list = buildTemplateTypes<T, Ts...>(); // Build type list
+    return fn_name + "<" + type_list + ">"; // Return function name with type list in angle brackets
+}
+
+// Function to load CUDA source code from a file
+std::string load_cuda_source(const std::string& filename) {
+    std::ifstream file(filename);
+    if (!file.is_open()) {
+        throw std::runtime_error("Failed to open file: " + filename);
+    }
+    std::ostringstream ss;
+    ss << file.rdbuf();
+    return ss.str();
+}
+
+/*
+Container class for the cached kernels. Provides functionality for launching compiled kernels as
+well as automatically resizing dynamic shared memory allocations, when needed. Kernels are compiled
+on first launch.
+*/
+class CachedKernel {
+
+  public:
+    CachedKernel(
+        std::string kernel_name,
+        std::string kernel_code,
+        std::string source_name,
+        std::vector<std::string> options
+    ) {
+        this->kernel_name = kernel_name;
+        this->kernel_code = kernel_code;
+        this->source_name = source_name;
+        this->options = options;
+    }
+
+    CachedKernel() = default;
+
+    void set(CUmodule m, CUfunction f, CUcontext c) {
+        this->module = m;
+        this->function = f;
+        this->context = c;
+    }
```

Comment:
This feels weird as a public function. Should it be removed and inlined where used?

Reply:
I've removed it.